### PR TITLE
start simplifying index code and make exclusive work

### PR DESCRIPTION
**Pull request by phritz**
separate logic commits:
- make exclusive work
- remove scan key index (offset)
- encode index keys ourselves and simplify IndexKey

TODO:
- eliminate ScanOptionsInternal and friends
- flatten ScanOptions for convenience

This feels a lot better to me than what we had before. @arv this changes the interface but i would wait for me to land next set of changes before doing the work to switch over. I'm going to simplify ScanOptions a bit.

Progress towards #250

### CODE REVIEW

File: src/db/index.rs [hunk 1]
```diff
@@ -187,80 +153,209 @@ fn get_index_keys(
...
 // Returns bytes that can be used to scan for the given secondary index value. These bytes
-// are different from the full index key bytes in that we do *not* want the trailing null
-// byte that bytekey appends to the secondary index value ("key\0" is not a prefix of "keyfoo").
-pub fn scan_key(secondary: &str) -> Result<Vec<u8>, GetIndexKeysError> {
-    let mut k = index_key(IndexValue::Str(secondary), &[])?;
-    k.truncate(k.len() - 1);
+// are different from the full index key bytes in that we do *not* want the null
+// byte separating secondary and primary keys ("key\0" is not a prefix of "keyfoo").
+// Exclusive scans are implemented by appending a 0x01 byte to the end of the scan key,
+// making it start atthe next higher value.
```
Comment (arv) [inline src/db/index.rs:198]: typo

File: src/db/index.rs [hunk 2]
```diff
@@ -187,80 +153,209 @@ fn get_index_keys(
...
+        k.push(0x01);
+    }
     Ok(k)
 }
 
+// Decodes an IndexKey encoded by encode_index_key, returning the values as a tuple
+// since IndexKey holds references.
+// TODO use this to set key in scan results
+#[allow(dead_code)]
+pub fn decode_index_key(index_key: &[u8]) -> Result<(String, Vec<u8>), DecodeIndexKeyError> {
```
Comment (arv) [inline src/db/index.rs:217]: Can the primary_bytes be a slice into the index key data?

**Review by arv — COMMENTED** [review]
LGTM

File: src/db/index.rs (hunk 1 above)
Comment (phritz) [reply src/db/index.rs:198]: due to several comment overhauls this paragraph got nuked

File: src/db/index.rs (hunk 2 above)
Comment (phritz) [reply src/db/index.rs:217]: yes good point, done in https://github.com/rocicorp/repc/commit/56bfc8fe354d0d1477e8f653cc10ab3f30360c27

**Pull request merged**